File: app/src/main/java/com/example/aaron/recipeassistant/browserecipes/ui/RecipeRecyclerViewAdapter.java
package com.example.aaron.recipeassistant.browserecipes.ui;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.util.Pair;
import android.support.v7.widget.RecyclerView;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.example.aaron.recipeassistant.model.RecipeDTO;
import com.example.aaron.recipeassistant.model.RecipeList;
import com.example.aaron.recipeassistant.model.RecipeMapper;
import com.example.aaron.recipeassistant.model.Recipe;
import com.example.aaron.recipeassistant.R;
import com.example.aaron.recipeassistant.readrecipe.ReadRecipeActivity;
import com.squareup.picasso.Picasso;

public class RecipeRecyclerViewAdapter extends RecyclerView.Adapter<RecipeRecyclerViewAdapter.RecipeViewHolder> {

    private Context context;
    private Activity activity;
    private RecipeList recipeList;
    private int imageSize;
    private Picasso picasso;

    public RecipeRecyclerViewAdapter(@NonNull Activity activity, int columnCount) {
        this.context = activity;
        this.activity = activity;

        DisplayMetrics displayMetrics = new DisplayMetrics();
        activity.getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        imageSize = (displayMetrics.widthPixels / columnCount);
        picasso = Picasso.with(context);
        //picasso.setIndicatorsEnabled(true);
    }

    void swapMealList(RecipeList newRecipeList) {
        recipeList = newRecipeList;
        notifyDataSetChanged();
    }

    @Override
    public RecipeViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.recipe_list_item, parent, false);
        view.setFocusable(true);
        return new RecipeViewHolder(view);
    }

    @Override
    public void onBindViewHolder(RecipeViewHolder holder, int position) {
        RecipeDTO meal = recipeList.getMeals().get(position);
        holder.meal = meal;
        String mealName = meal.getStrMeal().trim().toUpperCase().replace(" ", "\n");
        holder.recipeName.setText(mealName);
        picasso.load(meal.getStrMealThumb())
                .into(holder.recipePhoto);


    }

    @Override
    public int getItemCount() {
        if (null == recipeList) return 0;
        return recipeList.getMeals().size();
    }

    class RecipeViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final ImageView recipePhoto;
        private final TextView recipeName;
        private RecipeDTO meal;

        public RecipeViewHolder(View itemView) {
            super(itemView);

            recipePhoto = (ImageView) itemView.findViewById(R.id.recipe_photo);
            recipeName = (TextView) itemView.findViewById(R.id.recipe_name);
            RelativeLayout.LayoutParams params = new RelativeLayout.LayoutParams(imageSize, imageSize);
            recipeName.setLayoutParams(params);
            recipePhoto.setLayoutParams(params);
            itemView.setOnClickListener(this);
            activity.startPostponedEnterTransition();
        }

        @Override
        public void onClick(View v) {
            RecipeMapper recipeMapper = new RecipeMapper(meal);
            Recipe recipe = recipeMapper.getRecipe();

            Intent intent = new Intent(context, ReadRecipeActivity.class);
            intent.putExtra("recipe", recipe);

            String transImageName = context.getString(R.string.trans_img);
            String transTitleName = context.getString(R.string.trans_title);

            Pair<View, String> transImage = Pair.create((View)recipePhoto, transImageName);
            //Pair<View, String> transTitle = Pair.create((View) recipeName, transTitleName);

            ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(activity, transImage);

            activity.startActivity(intent, options.toBundle());
        }
    }
}

File: app/src/main/java/com/example/aaron/recipeassistant/browserecipes/ui/RecipeRecyclerViewAdapter.kt
package com.example.aaron.recipeassistant.browserecipes.ui

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.support.v4.app.ActivityOptionsCompat
import android.support.v4.util.Pair
import android.support.v7.widget.RecyclerView
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView

import com.example.aaron.recipeassistant.R
import com.example.aaron.recipeassistant.model.Recipe
import com.example.aaron.recipeassistant.readrecipe.ReadRecipeActivity
import com.squareup.picasso.Picasso

class RecipeRecyclerViewAdapter internal constructor(activity: Activity, columnCount: Int) : RecyclerView.Adapter<RecipeRecyclerViewAdapter.RecipeViewHolder>() {

    private val context: Context
    private val activity: Activity
    private var recipeList: List<Recipe> = listOf()
    private val imageSize: Int
    private val picasso: Picasso

    init {
        this.context = activity
        this.activity = activity

        val displayMetrics = DisplayMetrics()
        activity.windowManager.defaultDisplay.getMetrics(displayMetrics)
        imageSize = displayMetrics.widthPixels / columnCount
        picasso = Picasso.with(context)
    }

    internal fun swapMealList(newRecipeList: List<Recipe>) {
        recipeList = newRecipeList
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecipeViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.recipe_list_item, parent, false)
        view.isFocusable = true
        return RecipeViewHolder(view)
    }

    override fun onBindViewHolder(holder: RecipeViewHolder, position: Int) {
        val recipe = recipeList[position]
        holder.recipe = recipe
        val mealName = recipe.title.trim { it <= ' ' }.toUpperCase().replace(" ", "\n")
        holder.recipeName.text = mealName
        picasso.load(recipe.imageUrl)
                .into(holder.recipePhoto)
    }

    override fun getItemCount() = recipeList.size

    inner class RecipeViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        val recipePhoto: ImageView = itemView.findViewById(R.id.recipe_photo)
        val recipeName: TextView = itemView.findViewById(R.id.recipe_name)
        lateinit var recipe: Recipe

        init {
            val params = RelativeLayout.LayoutParams(imageSize, imageSize)
            recipeName.layoutParams = params
            recipePhoto.layoutParams = params
            itemView.setOnClickListener(this)
            activity.startPostponedEnterTransition()
        }

        override fun onClick(v: View) {
            val intent = Intent(context, ReadRecipeActivity::class.java)
            intent.putExtra("recipe", recipe)

            val transImageName = context.getString(R.string.trans_img)
            val transImage = Pair.create(recipePhoto as View, transImageName)
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation(activity, transImage)
            activity.startActivity(intent, options.toBundle())
        }
    }
}
